File: solutions/erlang/grains/1/src/grains.erl
-module(grains).

-export([square/1, total/0]).

square(Square)
  when (Square < 1) or (Square > 64) ->
  {error, "square must be between 1 and 64"};

square(Square) ->
  round(math:pow(2, Square - 1)).

total() ->
  Squares = lists:seq(1, 64),
  Grains = lists:map(collect_grain(), Squares),
  lists:foldl(add_grain(), _Accumulator = 0, Grains).

add_grain() ->
  fun(Grain, Accumulator) ->
    Accumulator + Grain
  end.

collect_grain() ->
  fun(Square) ->
    square(Square)
  end.

File: solutions/elixir/grains/1/lib/grains.ex
defmodule Grains do
  @doc """
  Calculate two to the power of the input minus one.
  """
  @spec square(pos_integer) :: pos_integer
  def square(integer)
      when integer < 1 or integer > 64,
      do: {:error, "The requested square must be between 1 and 64 (inclusive)"}

  def square(integer),
    do: {:ok, round(:math.pow(2, integer - 1))}

  @doc """
  Adds square of each number from 1 to 64.
  """
  @spec total :: pos_integer
  def total() do
    total =
      1..64
      |> Enum.map(&square/1)
      |> Enum.reduce(_accumulator = 0, &add_grain/2)

    {:ok, total}
  end

  defp add_grain(grain, accumulator) do
    {:ok, grain_count} = grain
    accumulator + grain_count
  end
end
